llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -3566,13 +3574,21 @@ static SDValue emitConditionalComparison(SDValue LHS, SDValue RHS,
       Opcode = AArch64ISD::CCMN;
       RHS = DAG.getConstant(Imm.abs(), DL, Const->getValueType(0));
     }
-  } else if (RHS.getOpcode() == ISD::SUB) {
-    SDValue SubOp0 = RHS.getOperand(0);
-    if (isNullConstant(SubOp0) && (CC == ISD::SETEQ || CC == ISD::SETNE)) {
-      // See emitComparison() on why we can only do this for SETEQ and SETNE.
-      Opcode = AArch64ISD::CCMN;
-      RHS = RHS.getOperand(1);
-    }
+  } else if (isCMN(RHS, CC, DAG)) {
+    Opcode = AArch64ISD::CCMN;
+    RHS = RHS.getOperand(1);
+  } else if (isCMN(LHS, CC, DAG) &&
+             (isIntEqualitySetCC(CC) ||
+              (isUnsignedIntSetCC(CC) && DAG.isKnownNeverZero(RHS)) ||
+              (isSignedIntSetCC(CC) && cannotBeIntMin(RHS, DAG)))) {
```

Comment:
Yes I do!

Let me show you!

Reply:
https://alive2.llvm.org/ce/z/8YTkie

Reply:
Actually if you want I can remove that part, but I have proof it works

Reply:
You know, I will revert this part only because it is risky keeping it is even though it is proven and tested

Reply:
Proof: 

It is known that if a > b, then b < a,

it is known that if b < a, then -b > -a

For the degenerative cases of INT_MIN and 0 for signed and unsigned comparisons, this is where knownbits steps in: https://alive2.llvm.org/ce/z/8YTkie

Reply:
I reverted the changed part so we can approve this. I will split the enhancement for another PR.

Reply:
I think the and_ult_eq_s0s1_or case was mis-compiling because of
```
    // Swap LHS and RHS if it wasn't an equality comparison
    // So we don't have to worry about changing the CC
    // a < b -> -b < -a
    std::swap(LHS, RHS);
```
The condition needed to be updated too.

Reply:
I swapped it so that the condition didn't need to be updated.

Reply:
> I think the and_ult_eq_s0s1_or case was mis-compiling because of
> 
> ```
>     // Swap LHS and RHS if it wasn't an equality comparison
>     // So we don't have to worry about changing the CC
>     // a < b -> -b < -a
>     std::swap(LHS, RHS);
> ```
> 
> The condition needed to be updated too.

The condition does not need to be updated because:

a < b -> -a > -b -> -b < -a

Reply:
Negation of both sides flips the condition and the swap flips it back. I did this because the CC isn't used in this method except for testing for eligibility, and this is much more feasible than refactoring so that we recompute the condition codes and arm condition codes.